src/EditorFeatures/CSharpTest/CodeActions/ConvertLinq/ConvertForEachToLinqQueryTests.cs

```diff
@@ -0,0 +1,2842 @@
+using System.Linq;
+class C
+{
+    List<int> M(IEnumerable<int> nums)
+    {
+        List<int> list = new List<int>(), list1 = new List<int>();
+        [|foreach (int n1 in nums)
+        {
+            foreach (int n2 in nums)
+            {
+                list.Add(n1);
+            }
+        }|]
+
+        return list;
+    }
+}
+";
+            string output = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    List<int> M(IEnumerable<int> nums)
+    {
+        List<int> list = new List<int>(), list1 = new List<int>();
+        list.AddRange(from int n1 in nums
+                      from int n2 in nums
+                      select n1);
+        return list;
+    }
+}
+";
+
+            await TestInRegularAndScriptAsync(source, output);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertForEachToQuery)]
+        public async Task ToListAssignToParameter()
+        {
+            string source = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    List<int> M(IEnumerable<int> nums, List<int> list)
+    {
+        list = new List<int>();
+        [|foreach (int n1 in nums)
+        {
+            foreach (int n2 in nums)
+            {
+                list.Add(n1);
+            }
+        }|]
+
+        return list;
+    }
+}
+";
+            string output = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    List<int> M(IEnumerable<int> nums, List<int> list)
+    {
+        list = (from int n1 in nums
+                from int n2 in nums
+                select n1).ToList();
+        return list;
+    }
+}
+";
+
+            await TestInRegularAndScriptAsync(source, output);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertForEachToQuery)]
+        public async Task ToListToArrayElement()
+        {
+            string source = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    void M(IEnumerable<int> nums, List<int>[] lists)
+    {
+        [|foreach (int n1 in nums)
+        {
+            foreach (int n2 in nums)
+            {
+                lists[0].Add(n1);
+            }
+        }|]
+    }
+}
+";
+            string output = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    void M(IEnumerable<int> nums, List<int>[] lists)
+    {
+        lists[0].AddRange(from int n1 in nums
+                          from int n2 in nums
+                          select n1);
+    }
+}
+";
+            await TestInRegularAndScriptAsync(source, output);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertForEachToQuery)]
+        public async Task ToListToNewArrayElement()
+        {
+            string source = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    void M(IEnumerable<int> nums, List<int>[] lists)
+    {
+        lists[0] = new List<int>();
+        [|foreach (int n1 in nums)
```

Comment:
it could be nice if this refactoring actually separated to 3

convert foreach to linq
merge list decl with usage 
inline return or something like that.

and basically this refactoring chain these 3 different refactoring.

Reply:
totally agree. thanks! will consider this at [https://github.com/dotnet/roslyn/issues/27980](https://github.com/dotnet/roslyn/issues/27980)

---
In reply to: [195857163](https://github.com/dotnet/roslyn/pull/27130#discussion_r195857163) [](ancestors = 195857163)